src/zarr/testing/strategies.py

```diff
@@ -86,11 +89,51 @@ def safe_unicode_for_dtype(dtype: np.dtype[np.str_]) -> st.SearchStrategy[str]:
 # i.e. stores.examples() will always return the same object per Store class.
 # So we map a clear to reset the store.
 stores = st.builds(MemoryStore, st.just({})).map(lambda x: sync(x.clear()))
-compressors = st.sampled_from([None, "default"])
-zarr_formats: st.SearchStrategy[ZarrFormat] = st.sampled_from([2, 3])
+zarr_formats: st.SearchStrategy[ZarrFormat] = st.sampled_from([3, 2])
 array_shapes = npst.array_shapes(max_dims=4, min_side=0)
 
 
+@st.composite  # type: ignore[misc]
+def codecs(
+    draw: st.DrawFn,
+    *,
+    zarr_formats: st.SearchStrategy[Literal[2, 3]] = zarr_formats,
+    dtypes: st.SearchStrategy[np.dtype] | None = None,
+) -> Any:
+    zarr_format = draw(zarr_formats)
+    # we intentional don't parameterize over `level` or `clevel` to reduce the search space
+    zarr_codecs = st.one_of(
+        st.builds(zcodecs.ZstdCodec),
+        st.builds(
+            zcodecs.BloscCodec,
+            shuffle=st.builds(
+                zcodecs.BloscShuffle.from_int, num=st.integers(min_value=0, max_value=2)
+            ),
+        ),
+        st.builds(zcodecs.GzipCodec),
+        st.builds(zcodecs.Crc32cCodec),
+    )
+    num_codecs_v2 = st.one_of(
+        st.builds(numcodecs.Zlib),
+        st.builds(numcodecs.LZMA),
+        st.builds(numcodecs.Zstd),
+        st.builds(numcodecs.Zlib),
+    )
+    num_codecs_v3 = st.one_of(
+        st.builds(ncodecs.Blosc),
+        st.builds(ncodecs.LZMA),
+        # st.builds(ncodecs.PCodec),
+        # st.builds(ncodecs.ZFPY),
+    )
```

Comment:
This search space is quite large. How can we be strategic here? 

I think we should test the codec pipeline separately with many different array types. For the `arrays` strategy, we could restrict ourselves to a much more limited choice (e.g. choose one among [a codec from numcodecs, a codec from zarr codecs, or None])